src/Common/src/System/Security/Cryptography/KeyFormatHelper.cs

```diff
@@ -0,0 +1,620 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.Asn1;
+
+namespace System.Security.Cryptography
+{
+    internal static class KeyFormatHelper
+    {
+        internal delegate void KeyReader<TRet, TParsed>(in TParsed key, in AlgorithmIdentifierAsn algId, out TRet ret);
+
+        internal static unsafe void ReadSubjectPublicKeyInfo<TRet, TParsed>(
+            string[] validOids,
+            ReadOnlySpan<byte> source,
+            KeyReader<TRet, TParsed> keyReader,
+            out int bytesRead,
+            out TRet ret)
+        {
+            fixed (byte* ptr = &MemoryMarshal.GetReference(source))
+            {
+                using (MemoryManager<byte> manager = new PointerMemoryManager<byte>(ptr, source.Length))
+                {
+                    ReadSubjectPublicKeyInfo(validOids, manager.Memory, keyReader, out bytesRead, out ret);
+                }
+            }
+        }
+
+        internal static ReadOnlyMemory<byte> ReadSubjectPublicKeyInfo(
+            string[] validOids,
+            ReadOnlyMemory<byte> source,
+            out int bytesRead)
+        {
+            // X.509 SubjectPublicKeyInfo is described as DER.
+            SubjectPublicKeyInfoAsn spki =
+                AsnSerializer.Deserialize<SubjectPublicKeyInfoAsn>(source, AsnEncodingRules.DER, out int read);
+
+            if (Array.IndexOf(validOids, spki.Algorithm.Algorithm.Value) < 0)
+            {
+                throw new CryptographicException(SR.Cryptography_NotValidPublicOrPrivateKey);
+            }
+
+            bytesRead = read;
+            return spki.SubjectPublicKey;
+        }
+
+        private static void ReadSubjectPublicKeyInfo<TRet, TParsed>(
+            string[] validOids,
+            ReadOnlyMemory<byte> source,
+            KeyReader<TRet, TParsed> keyReader,
+            out int bytesRead,
+            out TRet ret)
+        {
+            // X.509 SubjectPublicKeyInfo is described as DER.
+            SubjectPublicKeyInfoAsn spki =
+                AsnSerializer.Deserialize<SubjectPublicKeyInfoAsn>(source, AsnEncodingRules.DER, out int read);
+
+            if (Array.IndexOf(validOids, spki.Algorithm.Algorithm.Value) < 0)
+            {
+                throw new CryptographicException(SR.Cryptography_NotValidPublicOrPrivateKey);
+            }
+
+            TParsed parsed;
+
+            if (typeof(TParsed) == typeof(ReadOnlyMemory<byte>))
+            {
+                ReadOnlyMemory<byte> tmp = spki.SubjectPublicKey;
+                parsed = (TParsed)(object)tmp;
```

Comment:
You should be able to use Unsafe.As rather than relying on the JIT to recognize / transform this to avoid the boxing.

Reply:
I had that, and had a build error that I wasn't allowed to use the Unsafe package from shared framework.  Theoretically there's a different Unsafe, which maybe is callable to things in the shared framework and not in System.Private.CoreLib; but it didn't seem worth hunting down.

Reply:
> had a build error that I wasn't allowed to use the Unsafe package from shared framework

We use Unsafe from inside netcoreapp all over the place.